src/binder/query/query_graph_label_analyzer.cpp

```diff
@@ -94,49 +95,231 @@ void QueryGraphLabelAnalyzer::pruneNode(const QueryGraph& graph, NodeExpression&
+
+    auto forwardPruningFunc = [&](table_id_t srcTableID, table_id_t dstTableID) {
+        return srcTableIDSet.contains(srcTableID) && dstTableIDSet.contains(dstTableID);
+    };
     std::vector<TableCatalogEntry*> prunedEntries;
-    if (rel.getDirectionType() == RelDirectionType::BOTH) {
-        table_id_set_t srcBoundTableIDSet;
-        table_id_set_t dstBoundTableIDSet;
-        for (auto entry : rel.getSrcNode()->getEntries()) {
-            srcBoundTableIDSet.insert(entry->getTableID());
+    for (auto& entry : relEntries) {
+        auto& relEntry = entry->constCast<RelTableCatalogEntry>();
+        auto srcTableID = relEntry.getSrcTableID();
+        auto dstTableID = relEntry.getDstTableID();
+        auto satisfyForwardPruning = forwardPruningFunc(srcTableID, dstTableID);
+        if (directionType == RelDirectionType::BOTH) {
+            if (satisfyForwardPruning ||
+                (dstTableIDSet.contains(srcTableID) && srcTableIDSet.contains(dstTableID))) {
+                prunedEntries.push_back(entry);
+            }
+        } else {
+            if (satisfyForwardPruning) {
+                prunedEntries.push_back(entry);
+            }
         }
-        for (auto entry : rel.getDstNode()->getEntries()) {
-            dstBoundTableIDSet.insert(entry->getTableID());
+    }
+    return prunedEntries;
+}
+
+table_id_set_t QueryGraphLabelAnalyzer::collectRelNodes(const RelDataDirection direction,
+    std::vector<TableCatalogEntry*> relEntries) const {
+    table_id_set_t nodeIDs;
+    for (const auto& entry : relEntries) {
+        const auto& relEntry = entry->constCast<RelTableCatalogEntry>();
+        if (direction == RelDataDirection::FWD) {
+            nodeIDs.insert(relEntry.getDstTableID());
+        } else if (direction == RelDataDirection::BWD) {
+            nodeIDs.insert(relEntry.getSrcTableID());
+        } else {
+            KU_UNREACHABLE;
         }
-        for (auto& entry : rel.getEntries()) {
-            auto& relEntry = entry->constCast<RelTableCatalogEntry>();
-            auto srcTableID = relEntry.getSrcTableID();
-            auto dstTableID = relEntry.getDstTableID();
-            if ((srcBoundTableIDSet.contains(srcTableID) &&
-                    dstBoundTableIDSet.contains(dstTableID)) ||
-                (dstBoundTableIDSet.contains(srcTableID) &&
-                    srcBoundTableIDSet.contains(dstTableID))) {
-                prunedEntries.push_back(entry);
+    }
+    return nodeIDs;
+}
+
+std::pair<std::vector<table_id_set_t>, std::vector<table_id_set_t>>
+QueryGraphLabelAnalyzer::pruneRecursiveRel(const std::vector<TableCatalogEntry*>& relEntries,
+    const table_id_set_t srcTableIDSet, const table_id_set_t dstTableIDSet, size_t lowerBound,
+    size_t upperBound, RelDirectionType relDirectionType) const {
+    // src-->[dst,[rels]]
+    std::unordered_map<table_id_t, std::unordered_map<table_id_t, table_id_vector_t>>
```

Comment:
Up to use but I would use `std::unrdered_map<table_id_t, STRUCT>`

and STRUCT contains
```
table_id_t dstTableID
std::vector<tabel_id_t> relTableIDs
```

because we always access by srcTableID instead of `src&dstTableID`